src/gui/DetailsWidget.cpp

```diff
@@ -263,23 +264,21 @@ void DetailsWidget::getSelectedGroup(Group* selectedGroup)
 
 void DetailsWidget::updateTotp()
 {
-    if (m_locked) {
-        m_timer->stop();
```

Comment:
@adolfogc so the problem was that the  `m_timer` here was null? Why not just `if (m_locked && m_timer)` instead? I'm just trying to understand why `m_timer` is no longer a member variable.

Reply:
@louib The bug is caused by `m_timer` being undefined at first, please note that it isn't initialized to `nullptr` by the constructor. I think the offending code is the following:

```cpp
void DetailsWidget::getSelectedEntry(Entry* selectedEntry) 
// ...
    if (m_currentEntry->hasTotp()) {
        m_ui->totpButton->show();
        updateTotp(); // <-- [BUG]
        // updateTotp/1 gets called before m_timer and m_entry have a valid reference

        m_step = m_currentEntry->totpStep();

        m_timer = new QTimer(this);
        connect(m_timer, SIGNAL(timeout()), this, SLOT(updateTotp()));
        m_timer->start(m_step * 10);
    }
// ...
}
```

At the moment, my reason for removing those lines was that, IMHO, they were doing nothing, since they affected just a single timer instance (the one associated with the last selected entry) of all the possible. If a previous entries were selected, a QTimer instance could still be active. This means there is another bug here. What happens if two entries have TOTP information. We select a first entry that has TOTP information, so a QTimer instance gets created and it will call `updateTotp/1` every time it timeouts.  We then select a second entry with TOTP information and another QTimer instance gets created, it will also call `updateTotp/1` when it timeouts periodically. The bug manifests itself when the timing is such that both instances timeout with an interval greater than 0 seconds between them, i.e., they don't coincide.

This can be fixed by keeping the `m_timer` instance and initializing it to `nullptr`

```cpp
DetailsWidget::DetailsWidget(QWidget* parent)
    : QWidget(parent)
    , m_ui(new Ui::DetailsWidget())
    , m_locked(false)
    , m_timer(nullptr)
    , m_currentEntry(nullptr)
    , m_currentGroup(nullptr)
    , m_attributesWidget(nullptr)
    , m_autotypeWidget(nullptr)
    , m_selectedTabEntry(0)
    , m_selectedTabGroup(0)
{
    // ...
}
```

Applying the following changes:

```cpp
void DetailsWidget::getSelectedEntry(Entry* selectedEntry) 
// ...
    if (m_currentEntry->hasTotp()) {
        m_step = m_currentEntry->totpStep();

        if(nullptr != m_timer) { // a previous timer exists
            m_timer->stop();  
        }

        m_timer = new QTimer(this);
        connect(m_timer, SIGNAL(timeout()), this, SLOT(updateTotp()));
        updateTotp();
        m_timer->start(m_step * 10);
        m_ui->totpButton->show();
    }
// ...
}
```

And also:

```cpp
void DetailsWidget::updateTotp()
{
    if (!m_locked) {
        QString totpCode = m_currentEntry->totp();
        QString firstHalf = totpCode.left(totpCode.size() / 2);
        QString secondHalf = totpCode.right(totpCode.size() / 2);
        m_ui->totpLabel->setText(firstHalf + " " + secondHalf);
    } else {
      if (nullptr != m_timer) { // <-- As per your suggestion
          m_timer->stop(); 
      }
    }
}
```

I'll push a commit with these changes.

Reply:
At first I was using 1 QTimer for each entry but this was giving me problems since if you have multiple timer acting on the same Widget they where updating the DetailsWidget at different time making it unreliable (as you said), so I switched to 1 timer as member of DetailsWidget but I think I've missed something when testing this